File: 130. Surrounded Regions.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solve = function(board) {
    var BFS = function(i, j)
    {
        let queue = [{i: i, j: j}];
        while (queue.length > 0)
        {
            let temp = queue.shift();
            let x = temp.i, y = temp.j;
            if (x < 0 || x >= board.length || y < 0 || y >= board[0].length || board[x][y] != 'O')
                continue;
            board[x][y] = '*';
            queue.push({i: x-1, j: y});
            queue.push({i: x+1, j: y});
            queue.push({i: x, j: y-1});
            queue.push({i: x, j: y+1});
        }
    };
    if (board.length > 0)
    {
        for (let i = 0; i < board.length; i++)
        {
            BFS(i, 0);
            BFS(i, board[0].length-1);
        }
        for (let j = 0; j < board[0].length; j++)
        {
            BFS(0, j);
            BFS(board.length-1, j);
        }
        for (let i = 0; i < board.length; i++)
        {
            for (let j = 0; j < board[0].length; j++)
            {
                if (board[i][j] == 'O')
                    board[i][j] = 'X';
                else if (board[i][j] == '*')
                    board[i][j] = 'O';
            }
        }
    }
};

File: 130. Surrounded Regions.py
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if len(board) > 0:
            for i in xrange(len(board)):
                self.DFS(i, 0, board)
                self.DFS(i, len(board[0])-1, board)
            for j in xrange(len(board[0])):
                self.DFS(0, j, board)
                self.DFS(len(board)-1, j, board)
            for i in xrange(len(board)):
                for j in xrange(len(board[0])):
                    if board[i][j] == 'O':
                        board[i][j] = 'X'
                    elif board[i][j] == '*':
                        board[i][j] = 'O'
        
    def BFS(self, i, j, board):
        queue = [(i,j)]
        while queue:
            (x, y) = queue.pop(0)
            if x < 0 or x >= len(board) or y < 0 or y >= len(board[0]) or board[x][y] != 'O':
                continue
            board[x][y] = '*'
            queue += [(x-1, y)]
            queue += [(x+1, y)]
            queue += [(x, y-1)]
            queue += [(x, y+1)]
